src/primitives/decode/segwit.rs

```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: MIT
+
+//! Decoding of Segregated Witness address encoded as bech32 strings as specified by [BIP-173] (and [BIP-350]).
+
+use core::{fmt, str};
+
+use crate::primitives::checksum::Checksum;
+use crate::primitives::decode::{ByteIter, Fe32Iter};
+use crate::primitives::gf32::Fe32;
+use crate::primitives::hrp::Hrp;
+use crate::write_err;
+
+/// An HRP string that has been parsed and had the checksum validated.
+///
+/// Pre-parsing an HRP string only checks validity of the characters, it does not validate the
+/// checksum in any way - to validate convert to [`Parsed`].
+#[derive(Debug)]
+pub struct Parsed<'s> {
+    /// The parsed HRP and data.
+    inner: super::Parsed<'s>,
+    /// The first byte of `self.inner.data`.
+    witness_version: u8,
+}
+
+impl<'s> Parsed<'s> {
+    /// Parses an HRP string, treating the first data character as a witness version.
+    ///
+    /// This version byte does not appear in the extracted binary data, but is covered
+    /// by the checksum. It can be accessed with [`Self::witness_version`] and is also
+    /// returned from this constructor as a convenience.
+    pub fn new<Ck: Checksum>(s: &'s str) -> Result<Self, Error> {
```

Comment:
In 5c9b549a3738cec82f633c14e274854b82edce45:

We might as well do the bech32/bech32m decision here rather than making the user choose a checksum (which they cannot do until they've learned the witness version).

Reply:
Awesome, this prompted me to add all the test vector integration tests I wrote ages ago to the PR!